linked_list/linked_list.py

```diff
@@ -9,71 +12,129 @@ def __init__(self, value, next_node = None):
 # Defines the singly linked list
 class LinkedList:
     def __init__(self):
-      self.head = None # keep the head private. Not accessible outside this class
+        self.head = None # keep the head private. Not accessible outside this class
 
     # returns the value in the first node
     # returns None if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def get_first(self):
-        pass
+        return self.head.value if self.head else None
 
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(self, value):
-        pass
+        new_node = Node(value)
+        new_node.next = self.head
+        self.head = new_node
+        return self.head.value
+
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(N)
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        cur = self.head
+        while cur is not None:
+            if cur.value == value:
+                return True
+            cur = cur.next
+        else:
+            return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(N)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        length = 0
+        cur = self.head
+        while cur is not None:
+            length += 1
+            cur = cur.next
+        return length         
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(N)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        cur = self.head
+        while cur is not None and index != 0:
+            index -= 1
+            cur = cur.next
+        if index == 0:
+            return cur.value
+        else:
+            return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(N)
+    # Space Complexity: O(1)
     def get_last(self):

```

Comment:
👍 This works, nice.  Just a note that if you maintain a `tail` attribute this can be done in O(1) time (so can `add_last`).